EOLib/PacketHandlers/Chat/GroupChatHandler.cs

```diff
@@ -1,55 +1,59 @@
 ﻿using AutomaticTypeMapper;
-using EOLib.Domain.Character;
 using EOLib.Domain.Chat;
 using EOLib.Domain.Login;
-using EOLib.Domain.Map;
 using EOLib.Domain.Notifiers;
+using EOLib.Domain.Party;
+using EOLib.Net.Handlers;
 using Moffat.EndlessOnline.SDK.Protocol.Net;
 using Moffat.EndlessOnline.SDK.Protocol.Net.Server;
 using System.Collections.Generic;
+using System.Linq;
 
 namespace EOLib.PacketHandlers.Chat
 {
     [AutoMappedType]
-    public class GroupChatHandler : PlayerChatByIDHandler<TalkOpenServerPacket>
+    public class GroupChatHandler : InGameOnlyPacketHandler<TalkOpenServerPacket>
     {
         private readonly IChatRepository _chatRepository;
+        private readonly IPartyDataProvider _partyDataProvider;
         private readonly IEnumerable<IOtherCharacterEventNotifier> _otherCharacterEventNotifiers;
         private readonly IEnumerable<IChatEventNotifier> _chatEventNotifiers;
 
+        public override PacketFamily Family => PacketFamily.Talk;
         public override PacketAction Action => PacketAction.Open;
 
         public GroupChatHandler(IPlayerInfoProvider playerInfoProvider,
-                                ICurrentMapStateRepository currentMapStateRepository,
-                                ICharacterProvider characterProvider,
                                 IChatRepository chatRepository,
+                                IPartyDataProvider partyDataProvider,
                                 IEnumerable<IOtherCharacterEventNotifier> otherCharacterEventNotifiers,
                                 IEnumerable<IChatEventNotifier> chatEventNotifiers)
-            : base(playerInfoProvider, currentMapStateRepository, characterProvider)
+            : base(playerInfoProvider)
         {
             _chatRepository = chatRepository;
+            _partyDataProvider = partyDataProvider;
             _otherCharacterEventNotifiers = otherCharacterEventNotifiers;
             _chatEventNotifiers = chatEventNotifiers;
         }
 
         public override bool HandlePacket(TalkOpenServerPacket packet)
         {
-            return Handle(packet, packet.PlayerId);
-        }
+            var member = _partyDataProvider.Members.FirstOrDefault(member => member.CharacterID == packet.PlayerId);
```

Comment:
I've been trying to use optional pattern - `OptionalFirst(x => x.CharacterID == packet.PlayerId).MatchSome(x => ...)` - instead of linq's `XOrDefault` methods with comparison to `null`. More functional and all.

Reply:
Can you share an example of what you would do here? I'm not familiar with that I guess

Reply:
It's used all over the code base - if you search for `OptionalFirst` solution-wide you'll see the pattern.

There isn't much more to it than the example I gave. It's exactly like `SingleOrDefault` in that you pass a predicate to the method that provides a match. The result is an `Option<T>`, which is sort of like `OneOf` except it's either "Some" or "None". The chained call `MatchSome(x => ...)` does some action when there is some value.

It's used as a way to avoid the possibility of `NullReferenceException` being thrown due to a null value. It's a clear way to mark intent and force the caller to handle the return value regardless of whether there is one.

Reply:
@MrDanOak has fixed this now using `SingleOrNone` and `MatchSome` :)

Reply:
Looks good - if you guys sanity check this one more time to make sure the bugs in 352 are all no longer reproducible then I'm cool with merging this